### PR TITLE
Honor alternate root in system user/group provides

### PR DESCRIPTION
**Pull request by dmnks**
Commit 3617d160eb6a1a8a95689db5eb5648355ea60c2a missed the --root use case, causing it to only ever look up users and groups on the host, which is of course wrong. This is because rpmtsCheck() (where this dependency check is done) happens well before we enter the target chroot.

Fix that by setting the chroot just before the rpmug calls and adapting those functions to take the chroot path into account when constructing the final passwd/group file paths.

Ideally, we would enter the chroot already in rpmtsCheck() but that would cause --test transactions to require root privileges, which might (and likely would) break some existing use cases out there, so let's leave that consideration for the future (if at all) and just go with the "manual" string concatenation here.

Note that this new logic, much like the existing one, is based on the assumption that the rpmug cache is flushed across chroots. However, this is currently the case only when setting a new chroot, not when entering or leaving one, which will be fixed separately via #4093.

Add a test to cover the --root use case with system provides, too.

Fixes: #4094

### CODE REVIEW

**Comment by dmnks** [comment]
Oops, apparently I didn't run the tests locally with the sanitizers enabled :facepalm:

**Comment by pmatilai** [comment]
Been there, been there :smile:

**Comment by dmnks** [comment]
OK, memory leak fixed now :smile:

File: lib/depends.cc [hunk 1]
```diff
@@ -672,6 +673,8 @@ static int systemProvides(rpmts ts, rpmds dep)
     size_t nlen = strlen(n);
     char *dval = NULL;
 
+    rpmChrootSet(rpmtsRootDir(ts));
```
Comment (pmatilai) [inline lib/depends.cc]: I'd move these rpmChrootSet()'s to rpmtsCheck() level so they only get called once per the full dependency checking round. systemProvides() gets called once for every non-rpmlib() dependency which can be (tens of) thousands of times.

Other than that, seems fine to me.

Comment (dmnks) [reply lib/depends.cc]: Oh yep, initially I added them to rpmtsCheck() but then thought it could have some unintended consequences since there's a whole bunch of logic in that function. Except I didn't realize this would've been called repeatedly there, of course, oops :sweat_smile:

Will fix.

Comment (pmatilai) [reply lib/depends.cc]: Yeah, rpmChrootSet() does have some consequences because it (tries to) open the current directory, and that could fail in some funny situations that it hasn't been exposed so far. I guess we'll find out :sweat_smile:

Comment (dmnks) [reply lib/depends.cc]: Indeed, the worry pretty much boils down to that open() call, otherwise we're not using the rpmchroot API in rpmtsCheck(). One obvious use case would be when you, as a regular user, enter a directory only readable by the root user and then run a test transaction (still as a regular user), which would currently (probably) work. But it does seem like a corner case. Yet, it could break some existing scripts out there :sweat_smile:

**Comment by dmnks** [comment]
This needs a bit more tweaking, to address the comments above and also to make the subsequent #4085 into a smaller patch. I've got some of the pieces locally now and will put them together tomorrow. Converting to a draft for the time being.

**Comment by dmnks** [comment]
OK, made the following changes:
- Moved the chroot setup to `rpmtsCheck()` as agreed upon
- Changed the semantics of the new `rpmChrootPath()` slightly: return NULL if unset *or* when "/". This is going to be more useful in #4085 and also is more in line with the entry checks that we do in e.g. `rpmChrootIn()` where we consider "/" as a noop situation. In other words, `rpmChrootPath()` returns NULL if no chroot operation will be done.

File: lib/depends.cc [hunk 2]
```diff
@@ -1053,6 +1054,9 @@ int rpmtsCheck(rpmts ts)
     if (reqnotfilehash->empty())
 	reqnotfilehash = filedepHashFree(reqnotfilehash);
 
+    /* Look at target root for dependencies (currently just system provides) */
```
Comment (dmnks) [inline lib/depends.cc]: Now I realize this comment is not exactly obvious but I also wanted to avoid writing a long paragraph here...

Comment (pmatilai) [reply lib/depends.cc]: Hmm, that's indeed a bit misleading because it sounds as only system provides are being looked up from the root.

Comment (pmatilai) [reply lib/depends.cc]: Maybe something like "enable system provides lookup from the target root" instead?

Comment (dmnks) [reply lib/depends.cc]: I like that :+1:
